### ISSUE TITLE
Preservation of TypeRef tokens can cause invalid RIDs to be assigned during metadata importing phase

### ISSUE THREAD

**Issue by greenozon** [Washi1337/AsmResolver#329]
### AsmResolver Version

4.11.1

### .NET Version

4.7.2

### Operating System

Windows

### Describe the Bug

AsmResolver does not support .net images that are obfuscated and have some tricks in .NET tables

For example on the following .net binary that has 2 extra records in TypeRef .NET table

![image](https://user-images.githubusercontent.com/765343/177004078-022a3ead-23c1-4f4a-ad6b-fc3ed3c62970.png)

the code can't handle it and leaves as null, thus exception
![image](https://user-images.githubusercontent.com/765343/177004018-0726d56a-452b-43a6-bd03-22072c2870d2.png)

### How To Reproduce

just try to load assembly, eg:

```
				ManagedPEImageBuilder IMPEIB = new ManagedPEImageBuilder
				{
					DotNetDirectoryFactory = new DotNetDirectoryFactory
					{
						MetadataBuilderFlags = MetadataBuilderFlags.PreserveAll,
						MethodBodySerializer = new CilMethodBodySerializer
						{
							ComputeMaxStackOnBuildOverride = new bool?(false)
						}
					}
				};
				PEImageBuildResult IR = IMPEIB.CreateImage(this.AsmModule);
```

### Expected Behavior

assembly should be loaded smoothly

### Actual Behavior

DotNetDirectory attribute is not constructed

### Additional Context

Here is an example how this use-case solved in dnlib:

https://github.com/0xd4d/dnlib/blob/6f4dc6f0a7f86dbe12449b74d5501daf1192c187/src/DotNet/MD/TablesStream_Read.cs#L38

**Comment by Washi1337** [comment]
Could you please provide a sample binary for which this error occurs?

**Comment by greenozon** [comment]
yeah, here it is: https://www.sendspace.com/file/b7qmv4

**Comment by Washi1337** [comment]
This seems to be unrelated to the invalid type references that are present in the original metadata table. Rather, it is a more fundamental w.r.t. the preservation of metadata tokens. The following are initial bug analysis notes on the root cause:

`DiagnosticBag` in the build result shows a hint of the real underlying issue:

![Screenshot_20220703_212610](https://user-images.githubusercontent.com/3613449/177054471-a73bd110-141f-4e58-ad98-de0660db3a13.png)

In CFF Explorer, two references can be seen in the typeref table for `System.Environment` and `System.Environment+SpecialFolder`:

![Screenshot_20220703_212513](https://user-images.githubusercontent.com/3613449/177054431-3c1d57ae-feab-4d1e-ad06-8a3e324bf58c.png)

What's unusual about this, is that the nested type is defined _before_ the enclosing class in the type ref table. This is somewhat untypical for a normal C# compiler (which is probably the reason why this case was never tested), but is technically allowed according to spec (Note that this is not the case for nested type **definitions**, which **do** have to be defined after their enclosing types).

AsmResolver uses `DotNetDirectoryBuffer::AddTypeReference` to both add new type references, as well as import existing ones from the original metadata tables. The problem is that this function assumes that nested types are added _after_ the enclosing type, as it adds all declaring types recursively to the buffer, before adding the type that it was passed into as argument. When order is swapped like in the screenshot above, this results in types getting allocated the wrong RIDs. A similar problem can probably also be found for MemberRefParent's in `MemberReference`s. 

This will probably require some refactoring in the internals of the metadata buffers. I will see if I can get a fix up and running next week. This will probably be hotfixed in a v4.11.2.

**Comment by greenozon** [comment]
Thank you for the great analysis!
looking forward for making the SW even more bulletproof :)

**Comment by Washi1337** [comment]
Potential fix is in #330. Could you verify that this indeed solves your problem?

Note: If the assembly contains invalid metadata like those two invalid type refs in the end, make sure that configure AsmResolver in such a way that you are ignoring any reader errors. e.g.:

```csharp
var module = ModuleDefinition.FromFile(@"C:\file.exe", new ModuleReaderParameters(EmptyErrorListener.Instance));
```

**Comment by Washi1337** [comment]
You can download the latest packages from [AppVeyor](https://ci.appveyor.com/project/Washi1337/asmresolver/builds/44115807/artifacts).
